text/0000-rustdoc-json.md

```diff
@@ -0,0 +1,1124 @@
+be treated as opaque as they aren't guaranteed to be stable across compiler invocations. IDs are
+only valid/consistent within a single JSON blob. They cannot be used to resolve references between
+the JSON output of different crates (see [the Resolving IDs section](#resolving-ids)).
+
+## Crate
+
+A Crate is the root of the outputted JSON blob. It contains all doc-relevant information about the
+local crate, as well as some information about external items that are referred to locally.
+
+Name      | Type    | Description
+----------|---------|------------------------------------------------------------------------------
+`name`    | String  | The name of the crate. If `--crate-name` is not given, based on the filename.
+`version` | String  | (*Optional*) The version string given to `--crate-version`, if any.
+`includes_private`  | bool  | Whether or not the output includes private items.
+`root`    | [ID](#ID)      | The ID of the root module Item.
+`index`   | Map<[ID](#ID), [Item](#Item)> | A collection of all Items in the crate[\*](#resolving-ids).
+`paths`   | Map<[ID](#ID), [ItemSummary](#ItemSummary)> | Maps all IDs (even external ones[\*](#resolving-ids)) to a brief description including their name, crate of origin, and kind.
+`extern_crates` | Map<int, [ExternalCrate](#ExternalCrate)> | A map of "crate numbers" to metadata about that crate.
+
+### Resolving IDs
+
+The crate's `index` contains mostly local items, which includes impls of external traits on local
+types or local traits on external types. The exception to this is that external trait definitions
+and their associated items are also included in the `index` because this information is useful when
+generating the comprehensive list of methods for a type.
+
+This means that many IDs aren't included in the `index` (any reference to a struct, macro, etc.
+from a different crate). In these cases the fallback is to look up the ID in the crate's `paths`.
+That gives [enough information](#ItemSummary) about the item to create cross references or simply
+provide a name without copying all of the information about external items into the local
+crate's JSON output.
+
+### ExternalCrate
+
+Name      | Type    | Description
+----------|---------|------------------------------------------------------------------------------
+`name`    | String  | The name of the crate.
+`html_root_url` | String  | (*Optional*) The `html_root_url` for that crate if they specify one.
+
+### ItemSummary
+
+Name      | Type    | Description
+----------|---------|------------------------------------------------------------------------------
+`crate_num` | int   | A number corresponding to the crate this Item is from. Used as an key to the `extern_crates` map in [Crate](#Crate). A value of zero represents an Item from the local crate, any other number means that this Item is external.
+`path`    | [String] | The fully qualified path (e.g. `["std", "io", "lazy", "Lazy"]` for `std::io::lazy::Lazy`) of this Item.
```

Comment:
How will this work for generics in qualified paths (`std::vec::Vec<T>`)? Or for associated items (`std::vec::Vec::into_iter`, `<std::vec::Vec as IntoIterator>::IntoIter`? Or opaque types (`fn f() -> impl Iterator`)?

More things that could break this at [`QPath`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.QPath.html).

Reply:
These paths are only to the actual items corresponding to IDs, so if you saw `Vec<T>` in some function signature it'd give you the ID for `Vec` and when you look it up in `paths` you'd get the `ItemSummary` with `path = ["std", "vec", "Vec"]`. The info about the generic params would be contained at the use site (in the Type representation for `Vec<T>`). If it's a local type then you can look it up in `index` to get more info about what generic params it takes and their bounds, but for something like `Vec` you don't get that info in your crates json blob. You'd have to generate the json for `std` and go find it in there yourself (probably by using the path from `ItemSummary`).

For reference here's what you get in the example above (with most fields removed for brevity)

```rust
pub fn f<T>(arg: Vec<T>) { ... }
```

```json
{
  "name": "f",
  "kind": "function",
  "inner": {
    "decl": {
      "inputs": [
        [
          "arg",
          {
            "kind": "resolved_path",
            "inner": {
              "name": "Vec",
              "id": "5:4236",
              "args": {
                "angle_bracketed": {
                  "args": [ { "type": { "kind": "generic", "inner": "T" } } ],
                  "bindings": []
                }
              },
              "param_names": []
            }
          }
        ]
      ],
      "output": null
    },
    "generics": {
      "params": [
        { "name": "T", "kind": { "type": { "bounds": [] } } }
      ],
      "where_predicates": []
    }
  }
}

```

Reply:
Ok, that makes sense. How would you handle `<T as Trait>::AssociatedItem` though? Normalize the type before generating the path?

Reply:
In that case the "kind" for the type would be "qualified_path" which has [the necessary info](https://github.com/p1n3appl3/rfcs/blob/rustdoc-json/text/0000-rustdoc-json.md#kind--qualified_path).